src/dataflow/src/render/join/delta_join.rs

```diff
@@ -562,43 +513,35 @@ where
             let row_key = row_packer.finish_and_reuse();
             // Explicit drop to release borrow on `row` so that it can be returned.
             drop(datums_local);
-            Ok((row, row_key))
+            Ok((row_key, row, time))
         }
     });
 
     use differential_dataflow::AsCollection;
     use timely::dataflow::operators::OkErr;
 
     let mut datums = DatumVec::new();
-    let (oks, errs2) = dogsdogsdogs::operators::lookup_map(
+    let (oks, errs2) = dogsdogsdogs::operators::half_join(
         &updates,
         trace,
-        move |(_row, row_key), key| {
-            // Prefix key selector must populate `key` with key from prefix `row`.
-            key.clone_from(&row_key);
-        },
-        // TODO(mcsherry): consider `RefOrMut` in `lookup` interface to allow re-use.
-        move |(stream_row, _stream_row_key), diff1, lookup_row, diff2| {
+        |time| time.saturating_sub(1),
+        comparison,
+        // TODO(mcsherry): consider `RefOrMut` in `half_join` interface to allow re-use.
+        move |_key, stream_row, lookup_row| {
             let temp_storage = RowArena::new();
             let mut datums_local = datums.borrow();
             datums_local.extend(stream_row.iter());
             datums_local.extend(lookup_row.iter());
-            (
-                closure.apply(&mut datums_local, &temp_storage),
-                diff1 * diff2,
-            )
+            closure.apply(&mut datums_local, &temp_storage)
         },
-        // Three default values, for decoding keys into.
-        Row::pack::<_, Datum>(None),
-        Row::pack::<_, Datum>(None),
-        Row::pack::<_, Datum>(None),
     )
     .inner
     .ok_err(|(x, t, d)| {
         // TODO(mcsherry): consider `ok_err()` for `Collection`.
         match x {
-            Ok(x) => Ok((x, t, d)),
-            Err(x) => Err((DataflowError::from(x), t, d)),
+            (Ok(x), time) => Ok((x.map(|x| (x, time)), t, d)),
```

Comment:
`time` and `t` can have different values, right? Can these variable names be renamed so we can tell why `time` and `t` are different from each other?

Reply:
That's a good point. I'll clean that up. The "short version" is that `t` is held constant throughout this process, and `time` is allowed to vary upwards as interesting things happen to it. E.g. being joined with some time that is incomparable, or traversing a temporal filter, or .. maybe other things.

Reply:
I've put better names in place, but I put the explanation on the method rather than here.